core/win32/callback.c

```diff
@@ -4976,14 +4976,30 @@ check_internal_exception(dcontext_t *dcontext, CONTEXT *cxt,
         if (is_in_dynamo_dll((app_pc)pExcptRec->ExceptionAddress))
             is_DR_exception = true;
         else {
+            /*
+             * i#2144 : We exclude single step cases in a jump.
+             */
+            if (pExcptRec->ExceptionCode == EXCEPTION_SINGLE_STEP &&
+                !in_fcache(pExcptRec->ExceptionAddress)) {
+                fragment_t wrapper;
+                linkstub_t *l;
+                fragment_t *f = fragment_pclookup(dcontext, dcontext->next_tag, &wrapper);
```

Comment:
next_tag is only set on entering the cache.  This could have run thousands of blocks since then, so next_tag is irrelevant at this point.

It seems that to handle this we need to decode the stub and extract the linkstub pointer.  We have existing stub decoding for coarse stubs today but not for fine.